File: src/common/py/core/logging/formatter.py
import logging
import typing


class Formatter(logging.Formatter):
    """
    Customized log formatter.
    
    This formatter mainly colorizes log records for better readability.
    """
    _colors = {
        "time": 7,
        "scope": 28,
        "levels": {
            logging.DEBUG: 241,
            logging.INFO: 27,
            logging.WARNING: 214,
            logging.ERROR: 160,
        },
        "params": {
            "name": 93,
            "value": 126,
        },
    }
    
    def format(self, record: logging.LogRecord) -> str:
        """
        Colorizes the various portions of a log record.
        
        Returns:
            The formatted text.
        """
        scope = self._get_scope(record)
        params = [f"{self._color_wrap(k, self._colors['params']['name'], italic=True)}={self._color_wrap(str(v), self._colors['params']['value'], italic=True)}" for k, v in self._get_extra_params(record).items()]
        tokens = [
            self._color_wrap(self.formatTime(record, "%Y-%m-%d %H:%M:%S"), self._colors["time"]),
            " [",
            self._color_wrap(record.levelname, self._get_level_color(record.levelno), bold=True),
            "|" + self._color_wrap(scope, self._colors["scope"]) if scope != "" else "",
            "] ",
            record.msg,
            f" ({'; '.join(params)})" if len(params) > 0 else "",
        ]
        return "".join(tokens)
    
    def _color_wrap(self, text: str, fg_color: int = 0, bg_color: int | None = None, bold: bool = False, italic: bool = False) -> str:
        s: str = f"\x1b[38;5;{fg_color}m"
        if bold:
            s += "\x1b[1m"
        if italic:
            s += "\x1b[3m"
        if bg_color is not None:
            s += f"\x1b[48;5;{bg_color}m"
        return s + text + "\x1b[0m"
    
    def _get_level_color(self, level: int) -> int:
        return self._colors["levels"][level] if level in self._colors["levels"] else 0
    
    def _get_scope(self, record: logging.LogRecord) -> str:
        scope = ""
        if "scope" in record.__dict__ and record.scope is not None:
            scope = record.scope.lower()
        return scope

    def _get_extra_params(self, record: logging.LogRecord) -> typing.Dict[str, typing.Any]:
        params: typing.Dict[str, typing.Any] = {}
        if "extra_params" in record.__dict__ and record.extra_params is not None:
            params = record.extra_params
        return params

File: src/common/web/core/logging/Formatter.ts
import { LogRecord } from "./LogRecord";

export class Formatter {
    private static _colors = {
        "default": "#000000",
        "time": "#808080",
        "scope": "#008700",
        "levels": {
            "DEBUG": "#626262",
            "INFO": "#005FFF",
            "WARNING": "#FFAF00",
            "ERROR": "#D70000"
        },
        "params": {
            "name": "#8700FF",
            "value": "#AF0087"
        }
    };

    private _formattedText: string = "";
    private _stylesStack: string[] = [];

    public constructor(record: LogRecord) {
        this.format(record);
    }

    private format(record: LogRecord): void {
        let tokens = [
            this.colorWrap(this.getTimestampString(record), Formatter._colors.time),
            this.colorWrap(" [", Formatter._colors.default),
            this.colorWrapEx(record.levelName, this.getLevelColor(record.levelName), true, false),
            record.scope ? this.colorWrap("|", Formatter._colors.default) + this.colorWrap(record.scope, Formatter._colors.scope) : "",
            this.colorWrap("] ", Formatter._colors.default),
            this.colorWrap(record.msg, Formatter._colors.default),
            Object.entries(record.params).length > 0 ? this.colorWrap(" (", Formatter._colors.default) +
                this.getParametersList(record) + this.colorWrap(")", Formatter._colors.default) : ""
        ];

        this._formattedText = tokens.join("");
    }

    public get formattedText(): string {
        return this._formattedText;
    }

    public get stylesStack(): string[] {
        return this._stylesStack;
    }

    private colorWrap(text: string, color: string): string {
        return this.colorWrapEx(text, color, false, false);
    }

    private colorWrapEx(text: string, color: string, bold: boolean, italic: boolean): string {
        let styles = [`color: ${color};`];
        if (bold) {
            styles.push("font-weight: bold;");
        }
        if (italic) {
            styles.push("font-style: italic;");
        }
        this._stylesStack.push(styles.join(" "));
        return `%c${text}`;
    }

    private getLevelColor(levelName: string): string {
        if (levelName in Formatter._colors.levels) {
            return Formatter._colors.levels[levelName as (keyof typeof Formatter._colors.levels)];
        }
        return Formatter._colors.default;
    }

    private getParametersList(record: LogRecord): string {
        let paramsList: string[] = [];
        let objEntries = Object.entries(record.params);
        for (let i = 0; i < objEntries.length; ++i) {
            let [key, value] = objEntries[i];
            let tokens = [
                this.colorWrapEx(key, Formatter._colors.params.name, false, false),
                this.colorWrap("=", Formatter._colors.default),
                this.colorWrapEx(value, Formatter._colors.params.value, false, true),
            ];
            if (i < objEntries.length - 1) {
                tokens.push(this.colorWrap("; ", Formatter._colors.default));
            }
            paramsList.push(tokens.join(""));
        }
        return paramsList.join("");
    }

    private getTimestampString(record: LogRecord): string {
        let str = record.timestamp.toISOString().replace("T", " ");
        return str.slice(0, str.length - 5);
    }
}
